Cython/Utility/ModuleSetupCode.c

```diff
@@ -1588,15 +1588,28 @@ static int __Pyx_check_binary_version(void);
 /////////////// CheckBinaryVersion ///////////////
 
 static int __Pyx_check_binary_version(void) {
-    char ctversion[4], rtversion[4];
-    PyOS_snprintf(ctversion, 4, "%d.%d", PY_MAJOR_VERSION, PY_MINOR_VERSION);
-    PyOS_snprintf(rtversion, 4, "%s", Py_GetVersion());
-    if (ctversion[0] != rtversion[0] || ctversion[2] != rtversion[2]) {
+    char ctversion[5], rtversion[5] = {'\0'};
+    const char* rt_from_call = Py_GetVersion();
+    int same=1, i=0, found_dot=0;
+    PyOS_snprintf(ctversion, 5, "%d.%d", PY_MAJOR_VERSION, PY_MINOR_VERSION);
+    // slightly convoluted, but now we're into double digit version numbers we can no longer just rely on the length.
+    // Instead copy numbers and the first dot
+    for (; i<4 && ((*rt_from_call>='0' && *rt_from_call<='9') || (*rt_from_call=='.' && !found_dot)) ; ++i, ++rt_from_call) {
+        if (*rt_from_call=='.') found_dot = 1;
+        rtversion[i] = *rt_from_call;
+    }
```

Comment:
Let's untangle this a bit to make it more readable. Here's a suggestion.
```suggestion
    char ctversion[5], rtversion[5] = {'\0'};
    const char* rt_from_call = Py_GetVersion();
    int same, i, found_dot;
    PyOS_snprintf(ctversion, 5, "%d.%d", PY_MAJOR_VERSION, PY_MINOR_VERSION);
    // slightly convoluted, but now we're into double digit version numbers we can no longer just rely on the length.
    // Instead copy numbers and the first dot
    found_dot = 0
    for (i=0; i<4 && (*rt_from_call>='0' && *rt_from_call<='9') ; ++i) {
        if (*rt_from_call=='.') {
            if (found_dot) break;
            found_dot = 1;
        }
        rtversion[i] = *rt_from_call;
        ++rt_from_call;
    }
    same = 1;
```